libbeat/common/event.go

```diff
@@ -1,92 +1,223 @@
-			continue
-		case *MapStr:
-			v[key] = ConvertToGenericEvent(*value.(*MapStr))
-			continue
-		default:
+	return out, errs
+}
 
-			typ := reflect.TypeOf(value)
+// normalizeSlice normalizes each element of the slice and returns a []interface{}.
+func normalizeSlice(baseKey string, v reflect.Value) (interface{}, []error) {
+	var errs []error
+	var sliceValues []interface{}
 
-			if typ.Kind() == reflect.Ptr {
-				typ = typ.Elem()
-			}
+	n := v.Len()
+	for i := 0; i < n; i++ {
+		sliceValue, err := normalizeValue(joinKeys(baseKey, strconv.Itoa(i)), v.Index(i).Interface())
+		if len(err) > 0 {
+			errs = append(errs, err...)
+		}
+
+		sliceValues = append(sliceValues, sliceValue)
+	}
+
+	return sliceValues, errs
+}
+
+func normalizeValue(key string, value interface{}) (interface{}, []error) {
+	// Dereference pointers.
+	value = followPointer(value)
+
+	if value == nil {
+		return nil, nil
+	}
+
+	switch value.(type) {
+	case encoding.TextMarshaler:
+		text, err := value.(encoding.TextMarshaler).MarshalText()
+		if err != nil {
+			return nil, []error{errors.Wrapf(err, "key=%v: error converting %T to string", key, value)}
+		}
+		return string(text), nil
+	case bool, []bool:
+	case int, int8, int16, int32, int64:
+	case []int, []int8, []int16, []int32, []int64:
+	case uint, uint8, uint16, uint32, uint64:
+	case []uint, []uint8, []uint16, []uint32, []uint64:
+	case float32, float64:
+	case []float32, []float64:
+	case complex64, complex128:
+	case []complex64, []complex128:
+	case string, []string:
+	case Time, []Time:
+	case MapStr:
+		return normalizeMap(key, value.(MapStr))
+	case []MapStr:
+		return normalizeMapStrSlice(key, value.([]MapStr))
+	case map[string]interface{}:
+		return normalizeMap(key, value.(map[string]interface{}))
+	case []map[string]interface{}:
+		return normalizeMapStringSlice(key, value.([]map[string]interface{}))
+	default:
+		v := reflect.ValueOf(value)
 
-			switch typ.Kind() {
-			case reflect.Bool:
-			case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
-			case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
-			case reflect.Uintptr:
-			case reflect.Float32, reflect.Float64:
-			case reflect.Complex64, reflect.Complex128:
-			case reflect.String:
-			case reflect.UnsafePointer:
-			case reflect.Array, reflect.Slice:
-			//case reflect.Chan:
-			//case reflect.Func:
-			//case reflect.Interface:
-			case reflect.Map:
-				anothermap, err := MarshallUnmarshall(value)
-				if err != nil {
-					logp.Warn("fail to marshall & unmarshall map (%v): key=%v value=%#v",
-						key, value)
-					continue
-				}
-				v[key] = anothermap
-
-			case reflect.Struct:
-				anothermap, err := MarshallUnmarshall(value)
-				if err != nil {
-					logp.Warn("fail to marshall & unmarshall struct %v", key)
-					continue
-				}
-				v[key] = anothermap
-			default:
-				logp.Warn("unknown type %v", typ)
-				continue
+		switch v.Type().Kind() {
+		case reflect.Bool:
```

Comment:
In order to handle types like `type myAwesomeInt int32` reflection is used. Maybe be don't want to support `myAwesomeInt` in which case we could drop these case statements from the reflection block?